### PR TITLE
catch possible unnknown errors when loading submission

### PR DESCRIPTION
**Pull request by maikia**
It catches possible unknown errors when training the submission

hopefully closes: #471

### CODE REVIEW

File: ramp-engine/ramp_engine/tests/test_aws.py [hunk 1]
```diff
@@ -199,6 +199,33 @@ def test_launch_ec2_instances(boto_session_cls, use_spot_instance):
     launch_ec2_instances(config['worker'])
 
 
+@mock.patch('ramp_engine.aws.api._rsync')
+@mock.patch('ramp_engine.aws.api.launch_train')
+def test_aws_worker_launch_train_error(launch_train, test_rsync, caplog):
```
Comment (rth) [inline ramp-engine/ramp_engine/tests/test_aws.py:204]: Nice!

**Review by rth — APPROVED** [review]
Thanks @maikia ! LGTM though I'm not too closely acquainted with the AWS worker code base.

**Review by lucyleeow — COMMENTED** [review]
Looks good. Is the instance killed when worker status is 'error' ? This may be related to #474, as when the work status is error, we do this:
```Python
            if worker.status == 'error':
                set_submission_state(session, submission_id, 'checking_error')
                continue
```
which I don't think kills the instance so we have the 'ghost instances'.

**Comment by maikia** [comment]
> Looks good. Is the instance killed when worker status is 'error' ? This may be related to #474, as when the work status is error, we do this:
> 
> ```python
>             if worker.status == 'error':
>                 set_submission_state(session, submission_id, 'checking_error')
>                 continue
> ```
> 
> which I don't think kills the instance so we have the 'ghost instances'.

good point @lucyleeow 
What do you think now?

**Comment by lucyleeow** [comment]
This is good - with regards to the `teardown` when status is error, why did we not have this already for the condaworkers? I am not so familiar with the condaworkers - I am just wondering if there is/was any reason to not `teardown` in this place, before?

I also wonder if we also need to add a worker 'error status and teardown to the `collect_result` results part of the dispatcher. During this part the dispatcher is continuously checking if the worker has finished via `_wait_until_train_finished`, `_training_finished` and `_has_screen`. It seems possible to get error here as well - as you pointed out in #479

**Comment by maikia** [comment]
yes. You are right @lucyleeow . I was also wondering why teardown was not called here:
https://github.com/paris-saclay-cds/ramp-board/blob/26762a460d188033aea29144471b52aa06062727/ramp-engine/ramp_engine/dispatcher.py#L227

I couldn't find where it was called elsewhere when error status is set. 
Maybe here? (although it does not seem to answer it fully): https://github.com/paris-saclay-cds/ramp-board/blob/26762a460d188033aea29144471b52aa06062727/ramp-engine/ramp_engine/local.py#L126 

@glemaitre @rth could please confirm?

**Comment by glemaitre** [comment]
It depends if you want to do the action of the teardown which usually clean-up stuff.
If you just want to kill the worker but still be able to access the data that was computed, you probably don't want to call the teardown.

**Comment by maikia** [comment]
Thanks @glemaitre 
So do you agree on putting teardown here: for all the checking errors: https://github.com/paris-saclay-cds/ramp-board/blob/26762a460d188033aea29144471b52aa06062727/ramp-engine/ramp_engine/dispatcher.py#L171 
which are usually unknown errors coming from the worker itself.
Perhaps it is not an ideal long term solution and we should teardown or not depending on the exact error, but for now we are left with 'ghost instances' on AWS and I think this should solve the problem. But perhaps it could introduce another one if we are loosing the data on teardown?

File: ramp-engine/ramp_engine/dispatcher.py [hunk 2]
```diff
@@ -168,6 +168,11 @@ def launch_workers(self, session):
                 worker.status = 'error'
             if worker.status == 'error':
                 set_submission_state(session, submission_id, 'checking_error')
+                worker.teardown()  # kill the worker
+                self._logger.info(
+                            f'Worker {worker} killed due to an error '
+                            f'during training'
+                        )
```
Comment (glemaitre) [inline ramp-engine/ramp_engine/dispatcher.py:175]:
```suggestion
                self._logger.info(
                    f'Worker {worker} killed due to an error '
                    f'during training'
                )
```

**Comment by maikia** [comment]
Thanks @glemaitre @lucyleeow and @rth

**Pull request merged**